kuma/javascript/src/newsletter.jsx

```diff
@@ -89,76 +90,69 @@ export default function Newsletter() {
-            }).then(response => {
-                response.json().then(responseJSON => {
-                    if (responseJSON.success !== 'success') {
-                        let errorsArray = responseJSON.errors;
+        event.preventDefault();
+        let params = new URLSearchParams(
+            new FormData(newsletterForm)
+        ).toString();
+        fetch(newsletterSubscribeURL, {
+            method: 'POST',
+            headers: {
+                'X-Requested-With': 'XMLHttpRequest',
+                'Content-type': 'application/x-www-form-urlencoded'
+            },
+            body: params
+        })
+            .then(response => response.json())
+            .then(({ success, errors }) => {
+                if (success) {
+                    permanentlyHideNewsletter();
+                    setShowSuccessfulSubscription(true);
+                    ga('send', {
+                        hitType: 'event',
+                        eventCategory: 'newsletter',
+                        eventAction: 'progression',
+                        eventLabel: 'complete'
+                    });
+                    return;
+                }
 
-                        // if there are erros and the array contain at least one item
-                        if (errorsArray && errorsArray.length) {
-                            setErrorsArray(errorsArray);
-                            // if there was an error, but there are no items in the array
-                        } else if (errorsArray && errorsArray.length === 0) {
-                            if (newsletterForm) {
-                                // set the skip-fetch data attribute on the form
-                                newsletterForm.dataset.skipFetch = 'true';
-                                // and submit again for diagnoses on the server
-                                newsletterForm.submit();
-                            }
-                        } else {
-                            setShowNewsletter(false);
-                            permanentlyHideNewsletter();
-                            setShowSuccessfulSubscription(true);
-                            ga('send', {
-                                hitType: 'event',
-                                eventCategory: 'newsletter',
-                                eventAction: 'progression',
-                                eventLabel: 'complete'
-                            });
-                        }
-                    }
-                });
+                if (errors && errors.length) {
+                    setError(errors);
+                    // if there was an error, but there are no items in the array
+                } else if (errors && errors.length === 0) {
+                    // set the skip-fetch data attribute on the form
+                    newsletterForm.dataset.skipFetch = 'true';
+                    // and submit again for diagnoses on the server
+                    newsletterForm.submit();
```

Comment:
This feels clunky and I think you should consider avoiding the DOM API when it's not necessary. 
Make a closure function called something like...

```javascript
function postForm(form) {
    fetch(...).then(...).then(
         ...
         setSkipFetch(true)
         postForm(form)
    )
   
}
```
And on the form's `onSubmit` call that instead. E.g. 
```
<form onSubmit={event => {
  postForm(newsletterForm)
}}>
```

It just feels very jQuery to manually trigger things on the DOM instead of containing it inside actual JavaScript code.

Reply:
I think this might get a little awkward given how many state setters we'd have to pass in (or several callbacks). But I did take it as an incentive to look at extracting at least some of it and I think it does look better now.